kitsune/wiki/facets.py

```diff
@@ -73,62 +77,43 @@ def documents_for(locale, topics=None, products=None):
 def _documents_for(locale, topics=None, products=None):
     """Returns a list of articles that apply to passed in topics and products."""
     # First try to get the results from the cache
-    documents = cache.get(_documents_for_cache_key(locale, topics, products))
-    if documents:
+    cache_key = _cache_key(locale, topics, products)
+    documents_cache_key = f"documents_for:{cache_key}"
+    documents = cache.get(documents_cache_key)
+    if documents is not None:
         return documents
 
-    try:
-        # Then try ES
-        documents = _es_documents_for(locale, topics, products)
-        cache.add(_documents_for_cache_key(locale, topics, products), documents)
-    except TransportError:
-        # Finally, hit the database (through cache machine)
-        # NOTE: The documents will be the same ones returned by ES
-        # but they won't be in the correct sort (by votes in the last
-        # 30 days). It is better to return them in the wrong order
-        # than not to return them at all.
-        documents = _db_documents_for(locale, topics, products)
-        cache.add(_documents_for_cache_key(locale, topics, products), documents)
-
-    return documents
-
-
-def _es_documents_for(locale, topics=None, products=None):
-    """ES implementation of documents_for."""
-    s = (
-        DocumentMappingType.search()
-        .values_dict("id", "document_title", "url", "document_parent_id", "document_summary")
-        .filter(
-            document_locale=locale,
-            document_is_archived=False,
-            document_category__in=settings.IA_DEFAULT_CATEGORIES,
-        )
+    qs = (
+        Document.objects.filter(
+            locale=locale,
+            is_archived=False,
+            current_revision__isnull=False,
+            category__in=settings.IA_DEFAULT_CATEGORIES,
+        ).select_related("current_revision", "parent")
+        # speed up query by removing any ordering, since we're doing it in python:
+        .order_by()
     )
-
-    for topic in topics or []:
-        s = s.filter(topic=topic.slug)
-    for product in products or []:
-        s = s.filter(product=product.slug)
-
-    results = s.order_by("document_display_order", "-document_recent_helpful_votes")[:100]
-    results = DocumentMappingType.reshape(results)
-    return results
-
-
-def _db_documents_for(locale, topics=None, products=None):
-    """DB implementation of documents_for."""
-    qs = Document.objects.filter(
-        locale=locale,
-        is_archived=False,
-        current_revision__isnull=False,
-        category__in=settings.IA_DEFAULT_CATEGORIES,
-    ).prefetch_related("current_revision")
     for topic in topics or []:
-        qs = qs.filter(topics=topic)
+        qs = qs.filter(Q(topics=topic) | Q(parent__topics=topic))
```

Comment:
Let's rewrite this query without a `for` loop

Reply:
I've replaced this with `__in`, but of course, the result changes there (from filtering by documents in *all* topics, to filtering by documents in *any* topic), and the tests reflect that.

However, looking through this function's usage, I can't see any situation where it would make more sense to have *all* than *any* (especially when it comes to filtering by product). Shall I revert this change, or change the tests?

Reply:
I would say let's revert then or update this to filter on all elements. Do we need the filtering on parent topics here?